strutil/strutil.go

```diff
@@ -511,3 +512,37 @@ func Reverse(in string) string {
 	}
 	return string(out)
 }
+
+// ReplaceNonMatcher replaces all characters within a given that do not match
+// a given regular expression. The function returns the updated string and the
+// number of replacements
+func ReplaceNonMatcher(in, matcher string, replaceWith byte) (string, int) {
+	var adjustments int
+	work := []byte(in)
+	allowList := regexp.MustCompile(matcher)
+	for idx := range work {
```

Comment:
This function will break on many strings. The reason is that a string holds a sequence of bytes but since those bytes can contain unicode modifiers there is not a 1:1 mapping between a byte and an actual character in the string. As a result the right way to do this would be to keep the input string and range over it, which will give the starting indexes of the actual code points, as well as the rune itself. Then you should decide if the rune is what matches or doesn't match and rather than do an inline replacement use a string builder to build up the new string.

You can see the problems this creates at https://play.golang.com/p/3R-3kk-BO9I -- in the first example it iterates over the string and properly handles runes with multiple bytes. In the second example it properly iterates but by indexing you are seeing only the first unicode code point which is why the output is incorrect. In the third example, which matches what is happening in this function, you can see how it's not actually acting on the underlying string characters but rather a potentially partial representation of the character.

This matters for several reasons. One is that this is a generic string function package, and so while this may work for your use-case for user agents, it fails for general strings. Another is that this can actually return incorrect values because depending on the bytes that may make up a compound code point you may replace one byte of such a code point, leading the output string to have a changed value but not to the value you intended, at least not when treated as an actual UTF-8 string.

Reply:
@jefferai thanks a lot for this valuable feedback -- it's a blast! 

It's super interesting what you all got and I realize that I missed a lot of stuff.